### PR TITLE
chore: Add component specification

### LINKED ISSUE #8709: Define what `processed_bytes_total` means for sources and sinks
There have been some questions recently about what the `processed_bytes_total` metric means for a given component.

For example, in https://github.com/timberio/vector/pull/8692 we discussed whether we should report the metric before or after the HTTP request is sent.

A follow-up from this should be to ensure that all of our components report the metric in this fashion.

Open questions:

* Should this metric transition to `events_in_total` and `events_out_total` for all components. I think yes. Ref for transforms: #8405 
* Should this byte count include non-event parts of payloads? For example, the full HTTP request rather than just the body as we currently count it.

### LINKED ISSUE #8728: Create a Vector component specification
📢 It's high-time we create a specification for Vector components! 📢 All Vector components should adhere to a set of common behavior. Things like error handling, telemetry, and back pressure are crucial behavioral elements that must be consistent across component types. For example, issues #8693, #8368, and #8709 highlight inconsistencies in telemetry that make debugging and observability difficult. It's not enough to simply list the metrics that each component type should emit, we must be precise about what each metric means as evident by the inconsistencies in `bytes_out_total`.

This should serve as the initial work for #3900 so that we can align on the scope.

### PR DESCRIPTION
**Pull request by binarylogic**
Closes #8728
Closes #8709

This introduces the beginnings of a Vector component specification. The immediate need for this is to drive improvements around Vector's internal instrumentation. Issues like #8839, #8710, #8709, and #8693 all expose inconsistencies in Vector's internal instrumentation that presents problems for future work based on this data.

### CODE REVIEW

File: docs/specs/component.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+most sections will be broken along component lines for easy adherence.
+
+## Instrumentation
+
+### Events
+
+Vector implements an [event driven pattern ([RFC 2064]) for internal
+instrumentation. This section lists all required and optional events that a
+component MUST emit.
+
```
Comment (bruceg) [inline docs/specs/component.md:41]: Also worth noting that all errors should emit an `Event`, and it should increment an error counter (ie `processing_errors_total`) with tags.

**Review by bruceg — COMMENTED** [review]
This seems pretty minimal. I think it would be worth noting the standard tags logs and metrics are annotated with through the calling context. Should we also mandate counters on events/bytes out for transforms? There is also precedent for discarded events in some components. Some discussion of standardized attributes for logs would be useful as well (ex all error logs must have an `error` field, etc).

**Comment by binarylogic** [comment]
@bruceg this is the _very_ beginnings of the spec (hence the draft status). I wanted to get feedback on the structure before fleshing it out. I just wanted to make sure you were aware of that. I plan to add _many_ more events and metrics to this spec.

File: docs/specs/component.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+* The naming of the event MAY be component specific. For example,
+  `SocketEventReceived` since the `socket` source adds additional context.
+
+#### EventRecevied
+
+ALL components MUST emit an `EventReceived` event immediately after receiving
+a Vector event with the following telemetry:
+
+* Metrics
+   * MUST increment the `events_in_total` counter by 1.
```
Comment (jszwedko) [inline docs/specs/component.md:55]: Just noting we did discuss on a call that we batch these in places for performance reasons. For example: https://github.com/timberio/vector/pull/8383

File: docs/specs/component.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   * SHOULD increment the `bytes_in_total` counter by the total number of bytes.
+      * For sources, the total bytes coming off the wire.
+      * For everything else, the event's JSON byte size representation.
```
Comment (jszwedko) [inline docs/specs/component.md:58]: I think for sinks this should be the bytes out onto the wire.

There are some open questions about whether that should contain "wrapping" things like HTTP request headers on https://github.com/timberio/vector/issues/8709#issue-970550924

For transforms I like the idea of using the bytes size as calculated by https://github.com/timberio/vector/pull/8396 . I agree with you, as mentioned on the call, that we could make these metrics more specifically named, like: `event_bytes_in` and `event_bytes_out`. I guess `vector tap` would just collate both.

File: docs/specs/component.md (hunk 1 above)
Comment (jszwedko) [reply docs/specs/component.md:41]: Agreed, we'll want to discuss errors in this doc too.

**Review by jszwedko — COMMENTED** [review]
This seems like a good start to me. I like the direction it is heading.

File: docs/specs/component.md (hunk 1 above)
Comment (binarylogic) [reply docs/specs/component.md:42]: Yep, I'm on it.

File: docs/specs/component.md (hunk 2 above)
Comment (binarylogic) [reply docs/specs/component.md:74]: I call this out at the top of the section under implementation leeway.

File: docs/specs/component.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+#### Error
+
+*All components* MUST emit error events when an error occurs, and errors MUST be
+named with an `Error` suffix. For example, the `socket` source emits a
+`SocketReceiveError` representing any error that occurs while receiving data off
+of the socket.
+
+This specification does not list a standard set of errors that components must
+implement since errors are specific to the component.
```
[the remark this answers is not in the document]
Comment (binarylogic) [reply docs/specs/component.md:185]: Agree. @bruceg maybe you can do this as part of your observability work if patterns become obvious? If so, let's defer this until then.

**Review by jszwedko — APPROVED** [review]

**Pull request merged**